near-sdk/src/store/lookup_map/mod.rs

```diff
@@ -0,0 +1,533 @@
+    /// ```
+    /// use near_sdk::store::LookupMap;
+    ///
+    /// let mut count = LookupMap::new(b"m");
+    ///
+    /// for ch in [7, 2, 4, 7, 4, 1, 7] {
+    ///     let counter = count.entry(ch).or_insert(0);
+    ///     *counter += 1;
+    /// }
+    ///
+    /// assert_eq!(count[&4], 2);
+    /// assert_eq!(count[&7], 3);
+    /// assert_eq!(count[&1], 1);
+    /// assert_eq!(count.get(&8), None);
+    /// ```
+    pub fn entry(&mut self, key: K) -> Entry<K, V>
+    where
+        K: Clone,
+    {
+        let entry = self.get_mut_inner(&key);
+        if entry.value().is_some() {
+            // Value exists in cache and is `Some`
+            Entry::Occupied(OccupiedEntry { key, entry })
+        } else {
+            // Value exists in cache, but is `None`
+            Entry::Vacant(VacantEntry { key, entry })
+        }
+    }
+}
+
+impl<K, V, H> LookupMap<K, V, H>
+where
+    K: BorshSerialize + Ord,
+    V: BorshSerialize,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    /// Flushes the intermediate values of the map before this is called when the structure is
+    /// [`Drop`]ed. This will write all modified values to storage but keep all cached values
+    /// in memory.
+    pub fn flush(&mut self) {
+        let mut buf = Vec::new();
+        for (k, v) in self.cache.inner().iter_mut() {
+            if let Some(v) = v.get_mut() {
+                if v.is_modified() {
+                    let key = Self::lookup_key(&self.prefix, k);
+                    match v.value().as_ref() {
+                        Some(modified) => {
+                            buf.clear();
+                            BorshSerialize::serialize(modified, &mut buf)
+                                .unwrap_or_else(|_| env::panic_str(ERR_ELEMENT_SERIALIZATION));
+                            env::storage_write(&key, &buf);
+                        }
+                        None => {
+                            // Element was removed, clear the storage for the value
+                            env::storage_remove(&key);
+                        }
+                    }
+
+                    // Update state of flushed state as cached, to avoid duplicate writes/removes
+                    // while also keeping the cached values in memory.
+                    v.replace_state(EntryState::Cached);
+                }
+            }
+        }
+    }
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+#[cfg(test)]
+mod tests {
+    use super::LookupMap;
+    use crate::env;
+    use crate::hash::Keccak256;
+    use rand::seq::SliceRandom;
+    use rand::{Rng, SeedableRng};
+    use std::collections::HashMap;
+
+    #[test]
+    fn test_insert() {
+        let mut map = LookupMap::new(b"m");
+        let mut rng = rand_xorshift::XorShiftRng::seed_from_u64(0);
+        for _ in 0..500 {
+            let key = rng.gen::<u64>();
+            let value = rng.gen::<u64>();
+            map.insert(key, value);
+        }
+    }
```

Comment:
issue: randomized tests don't seem to be thorough. Specific problems: 

* `gen::<u64>` is guaranteed to generate different numbers, this'll never check the code path when the item is already there 
* we don't randomize the sequence of operation itself, it's the mixture of different operations that's interesting
* I think we want to keep these kinds of test as on of the main tool for checking sdk, so using `rand` directly feels like under-engineering. 

I've tried to come up with a better test here, and it found some problem. I haven't investigated the issue yet (it very well might be that my test is no good). Here's how I'd do this (using the astonishing `arbitrary` crate): 

```rust
use std::collections::HashMap;

use arbitrary::{Arbitrary, Unstructured};
use rand::seq::SliceRandom;
use rand::{Rng, RngCore, SeedableRng};

use crate::test_utils::test_env::setup_free;

#[derive(arbitrary::Arbitrary)]
enum Op {
    Add(u8, u8),
    Remove(u8),
    Flush,
    Restore,
    Get(u8),
}

#[test]
fn arbitrary() {
    setup_free();

    let mut rng = rand_xorshift::XorShiftRng::seed_from_u64(0);
    let mut buf = vec![0; 4096];
    for _ in 0..4096 {
        rng.fill_bytes(&mut buf);

        let mut lm = LookupMap::new(b"l");
        let mut hm = HashMap::new();
        let u = Unstructured::new(&buf);
        if let Ok(ops) = Vec::<Op>::arbitrary_take_rest(u) {
            for op in ops {
                match op {
                    Op::Add(k, v) => {
                        let r1 = lm.insert(k, v);
                        let r2 = hm.insert(k, v);
                        assert_eq!(r1, r2)
                    }
                    Op::Remove(k) => {
                        let r1 = lm.remove(&k);
                        let r2 = hm.remove(&k);
                        assert_eq!(r1, r2)
                    }
                    Op::Flush => {
                        lm.flush();
                    }
                    Op::Restore => {
                        lm = LookupMap::new(b"l");
                    }
                    Op::Get(k) => {
                        let r1 = lm.get(&k);
                        let r2 = hm.get(&k);
                        assert_eq!(r1, r2)
                    }
                }
            }
        }
    }
}
```

this currently gives

```
λ cargo test --release --package near-sdk --lib --features unstable -- store::lookup_map::tests::arbitrary --exact --nocapture
   Compiling near-sdk v4.0.0-pre.2 (/home/matklad/projects/near-sdk-rs/near-sdk)
    Finished release [optimized] target(s) in 2.75s
     Running unittests (target/release/deps/near_sdk-1734f3cd20b4f681)

running 1 test
thread 'store::lookup_map::tests::arbitrary' panicked at 'assertion failed: `(left == right)`
  left: `Some(147)`,
 right: `None`', near-sdk/src/store/lookup_map/mod.rs:409:29
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test store::lookup_map::tests::arbitrary ... FAILED
```

Reply:
Yeah these are just ported tests from the existing lookup map, I'll make sure this (and have plans to do the same for others) are fuzzed correctly

Reply:
So the error in the test shown above was due to the mocked storage not being cleared in-between iterations. Modifications of what's shown here added in [`be564e8` (#487)](https://github.com/near/near-sdk-rs/pull/487/commits/be564e8fafb26b59c49a2d1883776d25fc14ae73)